File: dotnet/examples/Pactman.NonprofitCheckPlus.Examples/Cases/Ex12FoundationClassification.cs
using System;
using System.Threading.Tasks;
using Pactman.NonprofitCheckPlus.Dev;

namespace Pactman.NonprofitCheckPlus.Examples.Cases;

/// <summary>
/// EX-12 — Organization type and foundation classification.
/// </summary>
/// <remarks>
/// Organization types, foundation and subsection classification for a grantmaker or DAF
/// display.
/// </remarks>
public sealed class Ex12FoundationClassification : IExample
{
    /// <inheritdoc />
    public string Id => "ex-12";

    /// <inheritdoc />
    public string Title => "Foundation and subsection classification, for a grantmaker display.";

    /// <inheritdoc />
    public async Task RunAsync(string[] args)
    {
        // A public charity and a private foundation side by side: the codes differ, and
        // the difference is what a DAF sponsor needs to see.
        using var context = ExampleContext.WithFixtures();

        foreach (var ein in new[] { Fixtures.Eins.PublicCharity, Fixtures.Eins.PrivateFoundation })
        {
            var result = await context.Client.Nonprofits.CheckAsync(ein);
            var organization = result.Nonprofit!;
            var bmf = organization.Bmf();

            Output.Heading(Output.Text(organization.OrganizationName));

            if (bmf is null)
            {
                Output.Bullet("No BMF data returned; there is nothing to classify.");

                continue;
            }

            Output.DisplayField(bmf, "subsection");
            Output.DisplayField(bmf, "subsection_description");
            Output.DisplayField(bmf, "foundation_code");
            Output.DisplayField(bmf, "foundation_code_description");
            Output.DisplayField(bmf, "foundation_type_code");
            Output.DisplayField(bmf, "foundation_type_description");
            Output.DisplayField(bmf, "foundation_509a_status");
            Output.DisplayField(bmf, "pf_filing_req_cd");

            var pub78 = organization.Pub78();

            Output.Field(
                "pub78 source_org_type_1",
                pub78 is null ? Output.NotReturned : Output.Display(pub78, "source_org_type_1"));

            foreach (var entry in organization.OrganizationTypes)
            {
                Output.Field("deductibility", $"{entry.DeductibilityStatusDescription} / {entry.DeductibilityLimitation}");
            }

            // The description fields are the ones to display. Codes are stable and
            // belong in storage; a description is what a reviewer can read.
            Output.Field("display as", bmf.FoundationTypeDescription ?? bmf.SubsectionDescription ?? "unclassified");

            // A DAF sponsor cares about this distinction, and the API states it — so
            // read the published field rather than inferring from the code.
            Output.Field("grant routing", bmf.FoundationTypeCode switch
            {
                "pc" => "public charity — ordinary DAF grant",
                "pf" => "private foundation — expenditure responsibility may apply",
                null => "unclassified — route to review",
                _ => $"code \"{bmf.FoundationTypeCode}\" — not one this application maps; route to review",
            });
        }

        Output.Note(
            "Store the codes, display the descriptions, and give any unmapped code a\n"
            + "route to review rather than a default. The IRS adds classifications.");
    }
}

File: ruby/examples/ex_12_foundation_classification.rb
# frozen_string_literal: true

# EX-12 — Organization type and foundation classification.
#
# A grantmaker or DAF needs the classification on screen: public charity or
# private foundation, which 509(a) paragraph, which deductibility limitation. The
# SDK maps every one of those fields and declares none of them grant-eligible.
#
# Note which values are read from the API's own `*_description` fields rather
# than a local table. Descriptions the source supplies stay correct when the
# source changes; a lookup table in your repository does not.
#
# Run:  PACTMAN_API_KEY=... bundle exec ruby examples/ex_12_foundation_classification.rb

require_relative "lib/fixture_api"
require_relative "lib/irs_codes"
require_relative "lib/print"

NCP = Pactman::NonprofitCheckPlus
include Print

# What a grant officer sees. Every value is copied, none is computed.
def classification_panel(nonprofit)
  bmf = NCP::Sources.bmf(nonprofit)
  first_type = NCP::Sources.pub78(nonprofit)&.organization_types&.first

  {
    "subsection code" => returned(bmf, :subsection),
    "subsection description" => returned(bmf, :subsection_description),
    "foundation code" => returned(bmf, :foundation_code),
    "foundation code description" => returned(bmf, :foundation_code_description),
    "foundation type code" => returned(bmf, :foundation_type_code),
    "foundation type description" => returned(bmf, :foundation_type_description),
    "509(a) status" => returned(bmf, :foundation_509a_status),
    "Pub 78 deductibility" =>
      IrsCodes.describe_deductibility_status(returned(first_type, :deductibility_status_description)).summary
  }
end

FixtureApi.with_fixture_api do |client|
  [FixtureApi::EINS[:public_charity], FixtureApi::EINS[:private_foundation]].each do |ein|
    nonprofit = client.nonprofits.check(ein).nonprofit

    if nonprofit.nil?
      puts "No record for #{ein}."
      next
    end

    heading "#{nonprofit.organization_name} (#{nonprofit.ein})"

    classification_panel(nonprofit).each { |label, value| field label, value }

    heading "  organization_types"

    types = NCP::Sources.pub78(nonprofit)&.organization_types

    if types.nil? || types.empty?
      puts "    none returned"
    else
      types.each_with_index do |entry, index|
        puts "    [#{index}] status=#{entry&.deductibility_status_description} " \
             "limitation=#{entry&.deductibility_limitation}"
      end
    end

    # A DAF's own rules live here, and they are visibly the DAF's. A private
    # foundation grantee is not disqualified — it is routed differently, because
    # expenditure responsibility and the deductibility limit both change.
    private_foundation = NCP::Sources.bmf(nonprofit)&.foundation_type_code == "pf"

    puts
    field "this application routes to",
          if private_foundation
            "private-foundation workflow — expenditure responsibility review"
          else
            "standard public-charity workflow"
          end
  end
end

note "Displaying a classification is not asserting grant eligibility. The SDK reports\n" \
     "the IRS classification; whether a grant may be made, and on what terms, is your\n" \
     "grantmaking policy and your counsel's call."
